File: frontend/src/lib/timezone.ts
/**
 * JST (Japan Standard Time) timezone utility functions for the frontend.
 * 
 * This module provides utilities for handling timezone conversions and getting 
 * current JST time to ensure consistent time handling across the frontend application.
 */

import { formatDistanceToNow, parseISO } from 'date-fns';
import { fromZonedTime, toZonedTime, formatInTimeZone } from 'date-fns-tz';
import { ja } from 'date-fns/locale';

// Japan Standard Time timezone
export const JST_TIMEZONE = 'Asia/Tokyo';

/**
 * Get current Date object in JST timezone.
 * 
 * @returns Date object representing current time in JST
 */
export function nowJST(): Date {
  return toZonedTime(new Date(), JST_TIMEZONE);
}

/**
 * Get current date string in JST timezone.
 * 
 * @returns Date string in YYYY-MM-DD format in JST
 */
export function todayJST(): string {
  return formatInTimeZone(new Date(), JST_TIMEZONE, 'yyyy-MM-dd');
}

/**
 * Get current time string in JST timezone.
 * 
 * @returns Time string in HH:mm:ss format in JST
 */
export function nowTimeJST(): string {
  return formatInTimeZone(new Date(), JST_TIMEZONE, 'HH:mm:ss');
}

/**
 * Convert Date object to JST timezone.
 * 
 * @param date - Date object to convert
 * @returns Date object converted to JST timezone
 */
export function toJST(date: Date): Date {
  return toZonedTime(date, JST_TIMEZONE);
}

/**
 * Convert JST time to UTC for API calls.
 * 
 * @param date - Date object in JST
 * @returns Date object converted to UTC
 */
export function jstToUTC(date: Date): Date {
  return fromZonedTime(date, JST_TIMEZONE);
}

/**
 * Format date in JST timezone with Japanese locale.
 * 
 * @param date - Date object or ISO string to format
 * @param formatStr - Format string for date formatting
 * @returns Formatted date string in JST with Japanese locale
 */
export function formatJST(date: Date | string, formatStr: string = 'yyyy年MM月dd日(E) HH:mm'): string {
  const dateObj = typeof date === 'string' ? parseISO(date) : date;
  return formatInTimeZone(dateObj, JST_TIMEZONE, formatStr, { locale: ja });
}

/**
 * Format date as display date (YYYY年MM月DD日(曜日)).
 * 
 * @param date - Date object or ISO string
 * @returns Formatted date string with Japanese day of week
 */
export function formatDateJST(date: Date | string): string {
  return formatJST(date, 'yyyy年MM月dd日(E)');
}

/**
 * Format time in JST timezone (HH:mm format).
 * 
 * @param date - Date object or ISO string
 * @returns Formatted time string in HH:mm format
 */
export function formatTimeJST(date: Date | string): string {
  return formatJST(date, 'HH:mm');
}

/**
 * Format time in JST timezone with seconds (HH:mm:ss format).
 * 
 * @param date - Date object or ISO string
 * @returns Formatted time string in HH:mm:ss format
 */
export function formatTimeWithSecondsJST(date: Date | string): string {
  return formatJST(date, 'HH:mm:ss');
}

/**
 * Format datetime in JST timezone for display.
 * 
 * @param date - Date object or ISO string
 * @returns Formatted datetime string
 */
export function formatDateTimeJST(date: Date | string): string {
  return formatJST(date, 'yyyy年MM月dd日(E) HH:mm');
}

/**
 * Get relative time from now in JST (e.g., "2時間前").
 * 
 * @param date - Date object or ISO string
 * @returns Relative time string in Japanese
 */
export function formatRelativeTimeJST(date: Date | string): string {
  const dateObj = typeof date === 'string' ? parseISO(date) : date;
  const jstDate = toJST(dateObj);
  
  return formatDistanceToNow(jstDate, { 
    addSuffix: true, 
    locale: ja,
    includeSeconds: true 
  });
}

/**
 * Check if two dates are the same day in JST.
 * 
 * @param date1 - First date
 * @param date2 - Second date
 * @returns True if both dates are on the same JST day
 */
export function isSameDayJST(date1: Date | string, date2: Date | string): boolean {
  const d1 = typeof date1 === 'string' ? parseISO(date1) : date1;
  const d2 = typeof date2 === 'string' ? parseISO(date2) : date2;
  
  const jst1 = formatInTimeZone(d1, JST_TIMEZONE, 'yyyy-MM-dd');
  const jst2 = formatInTimeZone(d2, JST_TIMEZONE, 'yyyy-MM-dd');
  
  return jst1 === jst2;
}

/**
 * Get start of day in JST (00:00:00).
 * 
 * @param date - Date object or ISO string (optional, defaults to current date)
 * @returns Date object representing start of day in JST
 */
export function getJSTStartOfDay(date?: Date | string): Date {
  const targetDate = date 
    ? (typeof date === 'string' ? parseISO(date) : date)
    : new Date();
  
  const jstDate = toJST(targetDate);
  const startOfDay = new Date(jstDate);
  startOfDay.setHours(0, 0, 0, 0);
  
  return startOfDay;
}

/**
 * Get end of day in JST (23:59:59.999).
 * 
 * @param date - Date object or ISO string (optional, defaults to current date)
 * @returns Date object representing end of day in JST
 */
export function getJSTEndOfDay(date?: Date | string): Date {
  const targetDate = date 
    ? (typeof date === 'string' ? parseISO(date) : date)
    : new Date();
  
  const jstDate = toJST(targetDate);
  const endOfDay = new Date(jstDate);
  endOfDay.setHours(23, 59, 59, 999);
  
  return endOfDay;
}

/**
 * Create a Date object from date and time strings in JST.
 * 
 * @param dateStr - Date string in YYYY-MM-DD format
 * @param timeStr - Time string in HH:mm or HH:mm:ss format
 * @returns Date object in JST
 */
export function createJSTDateTime(dateStr: string, timeStr: string): Date {
  const dateTimeStr = `${dateStr}T${timeStr}`;
  const date = new Date(dateTimeStr);
  return toZonedTime(date, JST_TIMEZONE);
}

/**
 * Get current month and year in JST.
 * 
 * @returns Object with current year and month in JST
 */
export function getCurrentMonthJST(): { year: number; month: number } {
  const now = nowJST();
  return {
    year: now.getFullYear(),
    month: now.getMonth() + 1 // JavaScript months are 0-indexed
  };
}

/**
 * Calculate elapsed time from start time to now in JST.
 * 
 * @param startTime - Start time string in HH:mm:ss format
 * @param currentDate - Current date string in YYYY-MM-DD format (optional)
 * @returns Elapsed time in seconds
 */
export function calculateElapsedTimeJST(startTime: string, currentDate?: string): number {
  const today = currentDate || todayJST();
  const startDateTime = createJSTDateTime(today, startTime);
  const now = nowJST();
  
  return Math.floor((now.getTime() - startDateTime.getTime()) / 1000);
}

/**
 * Format elapsed seconds to HH:mm:ss format.
 * 
 * @param seconds - Elapsed seconds
 * @returns Formatted time string
 */
export function formatElapsedTime(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const remainingSeconds = seconds % 60;
  
  return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${remainingSeconds.toString().padStart(2, '0')}`;
}

File: backend/app/utils/timezone.py
"""
JST (Japan Standard Time) timezone utility functions for the attendance management system.

This module provides utilities for handling timezone conversions and getting 
current JST time to ensure consistent time handling across the application.
"""

import pytz
from datetime import datetime, date, time
from typing import Optional

# Japan Standard Time timezone
JST = pytz.timezone('Asia/Tokyo')


def now_jst() -> datetime:
    """
    Get current datetime in JST timezone.
    
    Returns:
        datetime: Current datetime in JST timezone
    """
    return datetime.now(JST)


def today_jst() -> date:
    """
    Get current date in JST timezone.
    
    Returns:
        date: Current date in JST timezone
    """
    return now_jst().date()


def now_time_jst() -> time:
    """
    Get current time in JST timezone.
    
    Returns:
        time: Current time in JST timezone
    """
    return now_jst().time()


def to_jst(dt: datetime) -> datetime:
    """
    Convert datetime to JST timezone.
    
    Args:
        dt: Datetime object to convert
        
    Returns:
        datetime: Datetime converted to JST timezone
    """
    if dt.tzinfo is None:
        # If naive datetime, assume it's UTC
        dt = pytz.UTC.localize(dt)
    return dt.astimezone(JST)


def localize_jst(dt: datetime) -> datetime:
    """
    Localize naive datetime as JST.
    
    Args:
        dt: Naive datetime object
        
    Returns:
        datetime: Datetime localized to JST timezone
    """
    if dt.tzinfo is not None:
        raise ValueError("Datetime object is already timezone-aware")
    return JST.localize(dt)


def combine_date_time_jst(date_obj: date, time_obj: Optional[time] = None) -> datetime:
    """
    Combine date and time objects and localize to JST.
    
    Args:
        date_obj: Date object
        time_obj: Time object (if None, uses current JST time)
        
    Returns:
        datetime: Combined datetime localized to JST
    """
    if time_obj is None:
        time_obj = now_time_jst()
    
    # If time object has timezone info, remove it and use JST
    if hasattr(time_obj, 'tzinfo') and time_obj.tzinfo is not None:
        time_obj = time_obj.replace(tzinfo=None)
    
    naive_dt = datetime.combine(date_obj, time_obj)
    return JST.localize(naive_dt)


def format_jst_datetime(dt: datetime, format_str: str = "%Y-%m-%d %H:%M:%S %Z") -> str:
    """
    Format datetime in JST timezone.
    
    Args:
        dt: Datetime object
        format_str: Format string for datetime formatting
        
    Returns:
        str: Formatted datetime string in JST
    """
    jst_dt = to_jst(dt) if dt.tzinfo is not None else localize_jst(dt)
    return jst_dt.strftime(format_str)


def is_same_day_jst(dt1: datetime, dt2: datetime) -> bool:
    """
    Check if two datetimes are on the same day in JST.
    
    Args:
        dt1: First datetime
        dt2: Second datetime
        
    Returns:
        bool: True if both datetimes are on the same JST day
    """
    jst_dt1 = to_jst(dt1) if dt1.tzinfo is not None else localize_jst(dt1)
    jst_dt2 = to_jst(dt2) if dt2.tzinfo is not None else localize_jst(dt2)
    return jst_dt1.date() == jst_dt2.date()


def get_jst_start_of_day(date_obj: Optional[date] = None) -> datetime:
    """
    Get start of day (00:00:00) in JST for given date.
    
    Args:
        date_obj: Date object (if None, uses current JST date)
        
    Returns:
        datetime: Start of day in JST
    """
    if date_obj is None:
        date_obj = today_jst()
    return JST.localize(datetime.combine(date_obj, time.min))


def get_jst_end_of_day(date_obj: Optional[date] = None) -> datetime:
    """
    Get end of day (23:59:59.999999) in JST for given date.
    
    Args:
        date_obj: Date object (if None, uses current JST date)
        
    Returns:
        datetime: End of day in JST
    """
    if date_obj is None:
        date_obj = today_jst()
    return JST.localize(datetime.combine(date_obj, time.max))